public/app/features/templating/template_srv.ts

```diff
@@ -312,55 +318,95 @@ export class TemplateSrv implements BaseTemplateSrv {
     return scopedVar.value;
   }
 
-  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): string {
+  replace(target: string, scopedVars: ScopedVars, format: 'list'): string[];
+  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): string;
+  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): string | string[] {
     if (!target) {
       return target;
     }
+    switch (format) {
+      case 'list':
+        return TemplateSrv.cartesianProduct(
+          this.replaceWithoutFormatting(target, scopedVars).map(t => {
+            if (_.isArray(t.value)) {
+              return t.value.map(v => this.formatValue(v, t.format, t.variable));
+            } else {
+              return [this.formatValue(t.value, t.format, t.variable)];
+            }
+          })
+        ).map(t => t.join(''));
+      default:
+        return this.replaceWithoutFormatting(target, scopedVars, format)
+          .map(t => this.formatValue(t.value, t.format, t.variable))
+          .join('');
```

Comment:
not sure what is going on here, how can this work? formatValue just formats the value so in a expression like `my string $variable test` you cannot join the result of formatValue

Reply:
@torkelo [formatValue](https://github.com/grafana/grafana/pull/18812/files/615460f86a444aa1f9503ea9eb670cb488660da9#diff-ecab3b74c74dc34a9599591b1ab335a5R155) returns the formatted value, so I don't see why map would not work.

In the case "string $variable test", if $variable has values ["var1", "var2"] then the TargetToken objects returned by `replaceWithoutFormatting` would be:

1. {"value":"string "}
2. {"value":["var1","var2"]}
2. {"value":" test"}

This part is tested in [template_srv.test.ts](https://github.com/grafana/grafana/pull/18812/files/615460f86a444aa1f9503ea9eb670cb488660da9#diff-6819db476648351f2780cfdefec09421R293)

By the way @hugohaggmark, why was template_srv.test.ts removed in 00a9af00fc8a4c6d75dedd164da545e80c79a791?

Reply:
@pbaumard probably because many of those tests needed to be rewritten because we're using Redux state now. Thanks for finding that, I'll look into this asap.

Reply:
seems Torkel already added it back, so all is well again in the kingdom of variables :)